### ISSUE TITLE
Enhancement: Allow custom thresholding in filter wrappers

### ISSUE THREAD

**Issue by annette987** [mlr-org/mlr#2686]
Currently, makeFilterWrapper allows three different methods of thresholding to select the top scoring features:
- fw.perc - a percentage of the total number of features, 
- fw.abs - a fixed number of features, and
- fw.threshold - a threshold on the value produced by the filter. 

I would like to propose a fourth method to allow more flexibility. I suggest that makeFilterWrapper also accepts a custom thresholding function, which would allow users to do their own thresholding. For example, when using random forest variable importance, you might want to look for the biggest gap in importance values and only select the features with importance values above that gap.

In this proposal makeFilterWrapper would be extended to accept two new arguments:
- fw.func - the function itself, default value NULL
- fw.func.args - a named list of arguments to the function, default value NULL

The function would be passed a named list of arguments, the first of which would always be named "values" and would be a vector of values returned from the filter. Other arguments could be specified in fw.func.args.

The function would return the number of top-scoring features to select and would be mutually exclusive with fw.abs, fw.perc and fw.threshold. 

I have coded this and can do a PR if this idea is accepted.

**Comment by larskotthoff** [comment]
That sounds like a good idea, but I'm not sure if I understand entirely what you'd be able to do with that. What's the advantage of this over being able to provide a custom comparator function for sorting? Just want to avoid adding extra flexibility that is hard to maintain and test if we don't absolutely need it.

**Comment by annette987** [comment]
Where would you suggest a custom comparator function for sorting be provided? I am not clear on what you are suggesting, but understand your concerns.

The function I am suggesting is about more than just sorting. When you have a lot of variables (I am working with high-dimensional data), you can sort them in order of importance, but then you need to decide where to draw a line dividing those that are relevant and should be included in the final model from those that are redundant. At the moment the only option in mlr is to use a tuning wrapper to experiment with differing numbers of features (using fw.abs, fw.perc or fw.threshold) and select the number that gives the best model performance. This is a bit like trial and error. Providing a custom function that calculates the number of relevant features allows other, possibly more intuitive, methods to be used. For example the dividing line could be drawn where there is the the biggest difference between consecutive importance values when the features are sorted by importance.

I hope that makes more sense.

**Comment by larskotthoff** [comment]
Ok, I think I see what you mean. Could you provide an example of what this might look like please?

**Comment by annette987** [comment]
```
library(testthat)
library(survival)
library(mlr)

biggest_gap <- function(values, diff) {
  gap_size = 0
  gap_location = 0

  for (i in (diff+1):length(values)) {
    gap = values[[i - diff]] - values[[i]]
    if (gap > gap_size) {
      gap_size = gap
      gap_location = i - 1
    }
  }
  return(gap_location)
}

data(veteran)
set.seed(24601)
num_iters = 3
vet.task <- makeSurvTask(id = "VET", data = veteran, target = c("time", "status"))
vet.task <- createDummyFeatures(vet.task)
inner = makeResampleDesc("CV", iters = num_iters, stratify = TRUE)  # Tuning

filt1 = 
  makeFilterWrapper(
    makeLearner(cl = "surv.coxph", id = "coxph", predict.type = "response"), 
    fw.method = "randomForestSRC_importance",
    fw.func = biggest_gap,
    fw.func.args = list("diff" = 1),
    more.args = list("randomForestSRC_importance" = list(ntree = 50))
  )
r = resample(filt1, vet.task, inner, extract = function(model) {
  getFilteredFeatures(model)
})
expect_equal(r$extract, list("karno", "karno", "karno"))

```
I hope that helps.

**Comment by larskotthoff** [comment]
Thanks, that makes perfect sense -- looking forward to your PR!

**Comment by pat-s** [comment]
Thanks for contributing!

We can add support for a custom function if you already have a working implementation.

I'd suggest to add an argument named `fw.fun` which accepts a list:

First entry: the function
Second entry: further arguments (given in a list)

The function should be at least able to accept a numeric vector as the first argument (similar as your function in the example).
I would only want one test for this (testing a toy function that accepts a numeric vector and does some dummy ranking).

This gives users most flexibility and does not come with more maintenance effort for us.

**Comment by pat-s** [comment]
@annette987 Thanks for contributing so much lately to the package! We know it costs more time to go through all of this here rather than keeping all custom stuff private on ones machine. I assume that most of your scripts will rely on mlr2 but - have you ever thought about putting time into getting all of this working for mlr3?

You are probably aware that mlr2 is in "maintenance only" mode and that there is already the `mlr3filters` pkg. Yes, it lacks ensemble filters still but it will be the package which should be used in the future.
I am somewhat in the some situation as you where parts of my workflow rely on features in mlr2 which are not yet (fully) available in mlr3 but only with joint forces we can have a quick transition :)

Also feel free to join our Slack channel if you want to further contribute/discuss in any way to/about the mlr-package suite :)

**Comment by annette987** [comment]
> 
> 
> Thanks for contributing!
> 
> We can add support for a custom function if you already have a working implementation.
> 
> I'd suggest to add an argument named `fw.fun` which accepts a list:
> 
> First entry: the function
> Second entry: further arguments (given in a list)
> 
> The function should be at least able to accept a numeric vector as the first argument (similar as your function in the example).
> I would only want one test for this (testing a toy function that accepts a numeric vector and does some dummy ranking).
> 
> This gives users most flexibility and does not come with more maintenance effort for us.

Sorry I'm confused. I have provided a working implementation in PR #2692 that passes all tests and meets your requirements above. Are you asking for changes to that PR?

**Comment by annette987** [comment]
> 
> 
> @annette987 Thanks for contributing so much lately to the package! We know it costs more time to go through all of this here rather than keeping all custom stuff private on ones machine. I assume that most of your scripts will rely on mlr2 but - have you ever thought about putting time into getting all of this working for mlr3?
> 
> You are probably aware that mlr2 is in "maintenance only" mode and that there is already the `mlr3filters` pkg. Yes, it lacks ensemble filters still but it will be the package which should be used in the future.
> I am somewhat in the some situation as you where parts of my workflow rely on features in mlr2 which are not yet (fully) available in mlr3 but only with joint forces we can have a quick transition :)
> 
> Also feel free to join our Slack channel if you want to further contribute/discuss in any way to/about the mlr-package suite :)

I have only looked briefly at mlr3 and having put a lot of effort into understanding and using mlr2 I was reluctant to switch over. What are the advantages of mlr3? My work is mainly in survival analysis - does mlr3 implement that?  This work is for a PhD so time is limited. 
How do I join the Slack channel?

**Comment by pat-s** [comment]
> Sorry I'm confused. I have provided a working implementation in PR #2692 that passes all tests and meets your requirements above. Are you asking for changes to that PR?

Ah no, I just haven't yet looked into #2692. 

> I have only looked briefly at mlr3 and having put a lot of effort into understanding and using mlr2 I was reluctant to switch over. What are the advantages of mlr3? My work is mainly in survival analysis - does mlr3 implement that? This work is for a PhD so time is limited.
> How do I join the Slack channel?

Yeah sure, this was just to inform you about all of mlr3 in case you were not aware. No need to switch over, mlr2 functionality will stick around quite a bit.
mlr3 is aiming to do everything "better", "faster" and in a more simple way.
Survival analysis will be supported via https://mlr3proba.mlr-org.com/.
You can find instructions how to join the Slack group in the README.

**Comment by annette987** [comment]
Thanks for the info. mlr3proba looks interesting. I will have a look at it and mlr3filters when I get a chance .

**Comment by pat-s** [comment]
mlr3filters is still lacking ensemble support as we still need to discuss how to handle caching in mlr3.

**Comment by annette987** [comment]
I think I need an invitation to join the Slack group.

**Comment by larskotthoff** [comment]
Can you provide your email address please so we can invite you?

**Comment by annette987** [comment]
Sure. It is a.spooner@unsw.edu.au. Thanks.

**Comment by larskotthoff** [comment]
Thanks, invitation sent.

**Comment by pat-s** [comment]
fixed in #2692
